vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+ * 1. a routing track in X-direction or Y-direction (CHANX or CHANY) 
+ * 2. an input or an output of a logic block (IPIN or OPIN) 
+ * 3. a virtual source or sink node (SOURCE or SINK), which are starting/ending points of routing trees.
+ *
+ * Edge
+ * ----
+ * Each edge represents a switch between routing resources, which could be
+ * 1. a multiplexer
+ * 2. a tri-state buffer
+ * The switch information are categorized in the rr_switch_inf of RRGraph class.
+ *
+ * Guidlines on using the RRGraph data structure 
+ * =============================================
+ *
+ * For those want to access data from RRGraph
+ * ------------------------------------------
+ * Please refer to the detailed comments on each public accessors
+ *
+ * For those want to build/modify a RRGraph
+ * -----------------------------------------
+ * Do NOT add a builder to this data structure!
+ * Builders should be kept as free functions that use the public mutators
+ * We suggest developers to create builders in separated C/C++ source files
+ * outside the rr_graph header and source files
+ *
+ * After build/modify a RRGraph, please do run a fundamental check, a public accessor.
+ * to ensure that your RRGraph does not include invalid nodes/edges/switches/segements 
+ * as well as connections.
+ * This is a must-do check! 
+ *
+ * Example: 
+ *    RRGraph rr_graph;
+ *    ... // Building RRGraph
+ *    rr_graph.check(); 
+ *
+ * Optionally, we strongly recommend developers to run an advance check in check_rr_graph()  
+ * This guarantees legal and routable RRGraph for VPR routers.
+ *
+ * Note that these are generic checking codes. 
+ * If your RRGraph is designed for a special application context, please develop your 
+ * check_rr_graph() function to guarantee its correctness!
+ *
+ * Do NOT modify the coordinate system for nodes, they are designed for downstream drawers and routers
+ *
+ * For those want to extend RRGraph data structure
+ * --------------------------------------------------------------------------
+ * Please avoid modifying any existing public/private accessors/mutators
+ * in order to keep a stable RRGraph object in the framework
+ * Developers may add more internal data to RRGraph as well as associate accessors/mutators
+ * Please update and comment on the added features properly to keep this data structure friendly to be extended.
+ *
+ * Try to keep your extension within only graph-related internal data to RRGraph.
+ * In other words, extension is necessary when the new node/edge attributes are needed.
+ * RRGraph should NOT include other data which are shared by other data structures outside.
+ * Instead, using indices to point to the outside data source instead of embedding to RRGraph
+ * For example: 
+ *   For any placement/routing cost related information, try to extend t_rr_indexed_data, but not RRGraph
+ *   For any placement/routing results, try to extend PlaceContext and RoutingContext, but not RRGraph
+ * 
```

Comment:
The rr-graph is the single largest data structure in VPR, so avoid adding unnecessary information per node or per edge to it, as it will impact memory footprint.